### PR TITLE
MNT: hist2d now uses pcolormesh instead of pcolorfast

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

This is a redo of #4625 which simply changes hist2d to use pcolormesh instead of pcolorfast.  
The reason for doing so is to be able to handle log scales, which pcolormesh does but pcolorfast does not.

## PR Checklist

- [x] Code is PEP 8 compliant
- [x] Documentation is sphinx and numpydoc compliant
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Originally by @jankoeh...

**Comment by jklymak** [comment]
Difference between `pcolormesh` and `pcolorfast` is that `pcolormesh` has the anti-aliasing artifacts in PDF at the boundaries, whereas `pcolorfast` does not.  

In the conversation of #4625 folks seemed in favour of using `pcolormesh`, but I never quite saw the rationale.  Happy to close this PR if there is no strong rationale.

File: doc/api/api_changes/2017-12-12-JMK.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+`~.hist2d` now uses `~.pcolormesh` instead of `~.pcolorfast`
+------------------------------------------------------------
+
+`~.hist2d` now uses `~.pcolormesh` instead of `~.pcolorfast`.
```
Comment (anntzer) [inline doc/api/api_changes/2017-12-12-JMK.rst]: explicitly mention that this changes the return type?

Comment (efiring) [reply doc/api/api_changes/2017-12-12-JMK.rst]: And that the rationale is to provide support for log axes.

**Comment by dstansby** [comment]
I'm happy to push stuff to your branch if you don't mind?

**Comment by jklymak** [comment]
I don’t mind at all.  But don’t feel obliged - I will get to it the next couple of days...

File: doc/api/api_changes/2017-12-12-JMK.rst (hunk 1 above)
Comment (tacaswell) [reply doc/api/api_changes/2017-12-12-JMK.rst]: But `AxesImage now behaves correctly on log scales.

Comment (jklymak) [reply doc/api/api_changes/2017-12-12-JMK.rst]: OK, so we don't need this anymore? I didn't check recently that log scales now work.  

OTOH, Somewhat prefer pcolormesh, but thats just out of familiarity.

**Comment by efiring** [comment]
If the bins are not regularly spaced, the call to pcolorfast will return a PcolorImage, not an AxesImage, and PcolorImage does not support log scales.  There is no reason it couldn't.

**Comment by jklymak** [comment]
Ping @dstansby this is all done.  Not sure if @tacaswell is OK w/ the change, but I think @efiring makes a compelling argument....

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -1646,24 +1646,26 @@ def test_contour_colorbar():
...
+        remove_text=True, style='mpl20')
 def test_hist2d():
     np.random.seed(0)
     # make it not symmetric in case we switch x and y axis
     x = np.random.randn(100)*2+5
     y = np.random.randn(100)-2
     fig = plt.figure()
     ax = fig.add_subplot(111)
-    ax.hist2d(x, y, bins=10)
+    ax.hist2d(x, y, bins=10, rasterized=True)
```
Comment (dstansby) [inline lib/matplotlib/tests/test_axes.py]: Why the `rasterized=True` argument?

**Comment by jklymak** [comment]
Otherwise you get lines in the PDF output.

**Comment by dstansby** [comment]
🎉 thanks for this! (I'm plotting so many 2D histograms in log-space at the moment, so this will save me a lot of hassle!)

**Comment by jklymak** [comment]
NP.  Are we OK w/ the rasterized=True argument?  it doesnt' *really* matter, except it gets rid of the anti-aliasing artifact pcolormesh has for most PDF viewers.  https://github.com/jklymak/contourfIssues

**Pull request merged**